apps/homework/tasks.py

```diff
@@ -121,17 +142,32 @@ def post_submission(submission_pk, data_file=None):
                 # Temporary files can only be opened once, so both reads need to happen within this context manager
                 with data_to_upload.open() as fd:
                     numerator, denominator = retrieve_score_from_jupyter_notebook(data_to_upload, submission)
+                    # This seek happens, because the retrieve_score_from_jupyter_notebook reads the file.
+                    # The file position is reset here, so the next read, when the file is stored
+                    # (the .save method below), all of the file contents are read.
                     fd.seek(0)
                     submission.jupyter_notebook.save(data_to_upload.name, data_to_upload)
+
+                    submission_warnings = []
                     if denominator != submission.definition.jupyter_notebook_highest:
-                        # Add warning to submission model
-                        pass
+                        logger.warning(f'Score denominator is not equal to the one defined in homework definition. Warning in submission {submission.pk} for homework "{submission.definition.name}" (pk={submission.definition.pk}).')
+                        submission_warnings.append('Score denominator is not equal to the one defined in homework definition.')
                     if numerator > submission.definition.jupyter_notebook_highest:
                         # Add warning to submission model
                         numerator = submission.definition.jupyter_notebook_highest
+                        logger.warning(f'Score numerator greater than the maximum defined in homework definition. Warning in submission {submission.pk} for homework "{submission.definition.name}" (pk={submission.definition.pk}).')
+                        submission_warnings.append('Score numerator greater than the maximum defined in homework definition.')
                     if numerator < submission.definition.jupyter_notebook_lowest:
                         # Add warning to submission model
+                        logger.warning(f'Score numerator lower than the minimum defined in homework definition. Warning in submission {submission.pk} for homework "{submission.definition.name}" (pk={submission.definition.pk}).')
                         numerator = submission.definition.jupyter_notebook_lowest
+                        submission_warnings.append('Score numerator lower than the minimum defined in homework definition.')
+
+                    existing_messages = submission.reporting_messages
+                    if 'warnings' not in existing_messages:
+                        existing_messages['warnings'] = []
```

Comment:
the logic for setting this seems kinda weird? May be a case for `defaultdict` 
https://docs.python.org/3/library/collections.html#collections.defaultdict

I believe you can use this like...
```python
submission.reporting_messages = defaultdict(submission.reporting_messages or {}, {"warnings": []}, {"errors": []})
```

I may be misunderstanding something as well!

Reply:
Good call. I didn't know about `defaultdict`. It fits in well here. ✅